batch/batch.go

```diff
@@ -0,0 +1,96 @@
+package batch
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+// errSolo is a sentinel error indicating that the requester should re-run the
+// operation in isolation.
+var errSolo = errors.New(
+	"batch function returned an error and should be re-run solo",
+)
+
+type request struct {
+	*Request
+	errChan chan error
+}
+
+type batch struct {
+	db     kvdb.Backend
+	start  sync.Once
+	reqs   []*request
+	clear  func(b *batch)
+	locker sync.Locker
+}
+
+// trigger is the entry point for the batch and ensures that run is started at
+// most once.
+func (b *batch) trigger() {
+	b.start.Do(b.run)
+}
+
+// run executes the current batch of requests. If any individual requests fail
+// alongside others they will be retried by the caller.
+func (b *batch) run() {
+	// Clear the batch from its scheduler, ensuring that no new requests are
+	// added to this batch.
+	b.clear(b)
```

Comment:
I find this interleaving of batch and scheduler a bit confusing. Could we instead let the scheduler accumulate requests, when the timer ticks it will assemble the requests into a batch and trigger it?

Reply:
Still getting through this, but possibly couldn't a new batch instance be created that already removes the scheduled items from the scheduler? Seems like an odd "up call" in this context.

Reply:
fwiw this design is modeled off bbolt's batch/scheduler. the notation looks a little strange, but the `clear` method is a reference to a member function of the scheduler. the batch `b` is passed in so that the scheduler can determine whether this batch matches the scheduler's current batch, which we do because the execution order of two batches is undefined

Reply:
Not sure if we should copy bbolt if it is not to our standards 😛  Just looks like this could be made much simpler, but also not a blocker for me.

Reply:
not seeing why it doesn't meet our standards, how would you suggest making it simpler?

Reply:
See my first comment. I don't think it's a blocker, just saying that we could make this code more readable by not model the design off bbolt.